mkosi/__init__.py

```diff
@@ -1405,50 +1424,51 @@ def print_summary(config: MkosiConfig) -> None:
+    say("         Install Directory:", none_to_none(config.install_dir))
+    say("            Build Packages:", line_join_list(config.build_packages))
 
-    print("              Build Script:", path_or_none(config.build_script, check_script_input))
+    say("              Build Script:", path_or_none(config.build_script, check_script_input))
 
     env = [f"{k}={v}" for k, v in config.environment.items()]
     if config.build_script:
-        print("                 Run tests:", yes_no(config.with_tests))
+        say("                 Run tests:", yes_no(config.with_tests))
 
-    print("        Postinstall Script:", path_or_none(config.postinst_script, check_script_input))
-    print("            Prepare Script:", path_or_none(config.prepare_script, check_script_input))
-    print("           Finalize Script:", path_or_none(config.finalize_script, check_script_input))
+    say("        Postinstall Script:", path_or_none(config.postinst_script, check_script_input))
+    say("            Prepare Script:", path_or_none(config.prepare_script, check_script_input))
+    say("           Finalize Script:", path_or_none(config.finalize_script, check_script_input))
 
-    print("        Script Environment:", line_join_list(env))
-    print("      Scripts with network:", yes_no(config.with_network))
-    print("           nspawn Settings:", none_to_none(config.nspawn_settings))
+    say("        Script Environment:", line_join_list(env))
+    say("      Scripts with network:", yes_no(config.with_network))
+    say("           nspawn Settings:", none_to_none(config.nspawn_settings))
 
-    print("                  Password:", ("(default)" if config.password is None else "(set)"))
-    print("                 Autologin:", yes_no(config.autologin))
+    say("                  Password:", ("(default)" if config.password is None else "(set)"))
+    say("                 Autologin:", yes_no(config.autologin))
 
     if config.output_format == OutputFormat.disk:
-        print("\nVALIDATION:")
+        say(f"\n{b}VALIDATION{e}:")
 
-        print("                  Checksum:", yes_no(config.checksum))
-        print("                      Sign:", yes_no(config.sign))
-        print("                   GPG Key:", ("default" if config.key is None else config.key))
+        say("                  Checksum:", yes_no(config.checksum))
+        say("                      Sign:", yes_no(config.sign))
+        say("                   GPG Key:", ("default" if config.key is None else config.key))
 
-    print("\nHOST CONFIGURATION:")
+    say(f"\n{b}HOST CONFIGURATION{e}:")
 
-    print("        Extra search paths:", line_join_list(config.extra_search_paths))
-    print("      QEMU Extra Arguments:", line_join_list(config.qemu_args))
+    say("        Extra search paths:", line_join_list(config.extra_search_paths))
+    say("      QEMU Extra Arguments:", line_join_list(config.qemu_args))
+    page(f.getvalue(), config.pager)
```

Comment:
So this is probably moot, since `print` output is usually buffered anyway, but this does change the semantics of this somewhat, since lines are not printed individually anymore, but in one big block.

In principle `f` should also be closed after calling `.getvalue`, since this relies on garbage collection to deallocate the buffer.

Reply:
> So this is probably moot, since `print` output is usually buffered anyway, but this does change the semantics of this somewhat, since lines are not printed individually anymore, but in one big block.

Yes, that's true. But that's more-or-less what happens with a pager anyway. In practice the ability to have partial output is not interesting in this case. We generate the few dozen lines "immediately".



> In principle `f` should also be closed after calling `.getvalue`, since this relies on garbage collection to deallocate the buffer.

That's a common misconception. There is really no point in doing this. When using Python, we rely on "garbage collection to deallocate buffers" *all the time*. There is nothing special about this buffer to cause us to treat it in some special way.

Or in other words: I don't say `del f` at the end of the function. Likewise, I don't need to say `f.close()`. Both things happen automatically.

Of course, there are other situations where a file needs to be closed, or when a buffer needs to be deallocated explicitly, in particular when files are written to be read immediately, or buffers are large, but those are special cases which aren't relevant here.